library/aes.c

```diff
@@ -1523,35 +1523,90 @@ int mbedtls_aes_self_test( int verbose )
+                if( ret != 0 )
+                    goto exit;
+
+                if( offset != ( size_to_cipher % 16 ) )
+                {
+                    if( verbose != 0 )
+                        mbedtls_printf( "failed: calculated offset is not as expected\n" );
+                    ret = 1;
+                    goto exit;
+                }
+
+                /* resume the cipher, from offset */
+                ret = mbedtls_aes_crypt_ctr( &ctx, len - size_to_cipher, &offset, nonce_counter, stream_block,
+                                       buf + size_to_cipher, buf  + size_to_cipher );
+                if( ret != 0 )
+                    goto exit;
+
+                if( memcmp( buf, aes_test_ctr_pt[u], len ) != 0 )
+                {
+                    if( verbose != 0 )
+                        mbedtls_printf( "failed with cipher size %zu\n", size_to_cipher );
+
+                    ret = 1;
+                    goto exit;
+                }
+            }
         }
         else
         {
-            memcpy( buf, aes_test_ctr_pt[u], len );
-            aes_tests = aes_test_ctr_ct[u];
-        }
-
-        ret = mbedtls_aes_crypt_ctr( &ctx, len, &offset, nonce_counter,
-                                     stream_block, buf, buf );
-        if( ret != 0 )
-            goto exit;
-
-        if( memcmp( buf, aes_tests, len ) != 0 )
-        {
-            ret = 1;
-            goto exit;
+            len = aes_test_ctr_len[u];
+            for( size_to_cipher = 0; size_to_cipher < len; size_to_cipher++ )
+            {
+                memcpy( buf, aes_test_ctr_ct[u], len );
+                memcpy( nonce_counter, aes_test_ctr_nonce_counter[u], 16 );
+                offset = 0;
+
+                /*  cipher the first bytes with length size_to_cipher */
+                ret = mbedtls_aes_crypt_ctr( &ctx, size_to_cipher, &offset, nonce_counter, stream_block,
+                                        buf, buf );
+                if( ret != 0 )
+                    goto exit;
+
+                if( offset != ( size_to_cipher % 16 ) )
+                {
+                    if( verbose != 0 )
+                        mbedtls_printf( "failed: calculated offset is not as expected\n" );
+                    ret = 1;
+                    goto exit;
+                }
+
+                /* resume the cipher, from offset */
+                ret = mbedtls_aes_crypt_ctr( &ctx, len- size_to_cipher, &offset, nonce_counter, stream_block,
+                                       buf + size_to_cipher, buf + size_to_cipher );
+                if( ret != 0 )
+                    goto exit;
+                if( memcmp( buf, aes_test_ctr_pt[u], len ) != 0 )
+                {
+                    if( verbose != 0 )
+                        mbedtls_printf( "failed with cipher size %zu\n", size_to_cipher );
+
+                    ret = 1;
+                    goto exit;
+                }
+            }
```

Comment:
Please refactor the `if( mode == MBEDTLS_AES_DECRYPT )` blocks, as there is a lot of duplicated code and the differences between them should be clearer.